File: Assistant/lib/utility/storage_helper.dart
import '/common.dart';
import '/utility/finalizer.dart';
import '/utility/storage_path.dart';
import '/utility/convert_helper.dart';
import '/utility/platform_integration_manager.dart';
import 'dart:io' as lib;
import 'dart:typed_data' as lib;
import 'dart:convert' as lib;
import 'package:ffi/ffi.dart' as lib;
import 'package:win32/win32.dart' as lib;

// ----------------

enum StorageQueryType {
  userHome,
  applicationShared,
  applicationPersistent,
  applicationTemporary,
  applicationCache,
}

enum StoragePickType {
  loadFile,
  loadDirectory,
  saveFile,
}

// ----------------

class StorageHelper {

  // #region basic

  static Future<Boolean> exist(
    StoragePath target,
  ) async {
    if (target.type() != .absolute) {
      return false;
    }
    var targetString = target.emitNative();
    return await lib.FileSystemEntity.type(targetString, followLinks: false) != .notFound;
  }

  static Future<Void> copy(
    StoragePath target,
    StoragePath placement,
    Boolean     followLink,
  ) async {
    assertTest(await StorageHelper.exist(target));
    assertTest(!await StorageHelper.exist(placement));
    var placementParent = placement.parent();
    if (placementParent != null && !await StorageHelper.existDirectory(placementParent)) {
      await StorageHelper.createDirectory(placementParent);
    }
    var targetString = target.emitNative();
    var type = await lib.FileSystemEntity.type(targetString, followLinks: false);
    if (followLink && type == .link) {
      var referentType = await lib.FileSystemEntity.type(targetString, followLinks: true);
      if (referentType != .notFound) {
        type = referentType;
      }
    }
    if (type == .link) {
      var referent = await StorageHelper.resolveLink(target);
      var isDirectory = false;
      if (SystemChecker.isWindows) {
        var hTarget = targetString.toPcwstr();
        var attribute = lib.GetFileAttributes(hTarget);
        lib.calloc.free(hTarget);
        isDirectory = attribute.value != -1 && (attribute.value & lib.FILE_ATTRIBUTE_DIRECTORY) != 0;
      }
      if (SystemChecker.isLinux || SystemChecker.isMacintosh || SystemChecker.isAndroid || SystemChecker.isIphone) {
        isDirectory = false;
      }
      await StorageHelper.createLink(placement, referent, isDirectory);
    }
    else if (type == .file) {
      var data = await StorageHelper.readFileData(target);
      await StorageHelper.createFile(placement);
      await StorageHelper.writeFileData(placement, data);
    }
    else if (type == .directory) {
      await StorageHelper.createDirectory(placement);
      await for (var item in lib.Directory(targetString).list(recursive: false, followLinks: false)) {
        var itemName = item.path.substring(item.path.lastIndexOf(RegExp(r'[/\\]')) + 1);
        await StorageHelper.copy(target.join(itemName), placement.join(itemName), followLink);
      }
    }
    else {
      throw UnsupportedException();
    }
    return;
  }

  static Future<Void> rename(
    StoragePath target,
    StoragePath placement,
  ) async {
    assertTest(await StorageHelper.exist(target));
    assertTest(!await StorageHelper.exist(placement));
    var placementParent = placement.parent();
    if (placementParent != null && !await StorageHelper.existDirectory(placementParent)) {
      await StorageHelper.createDirectory(placementParent);
    }
    var targetString = target.emitNative();
    var placementString = placement.emitNative();
    var type = await lib.FileSystemEntity.type(targetString, followLinks: false);
    if (type == .link) {
      await lib.Link(targetString).rename(placementString);
    }
    else if (type == .file) {
      await lib.File(targetString).rename(placementString);
    }
    else if (type == .directory) {
      await lib.Directory(targetString).rename(placementString);
    }
    else {
      throw UnsupportedException();
    }
    return;
  }

  static Future<Void> remove(
    StoragePath target,
  ) async {
    assertTest(await StorageHelper.exist(target));
    var targetString = target.emitNative();
    var type = await lib.FileSystemEntity.type(targetString, followLinks: false);
    if (type == .link) {
      await lib.Link(targetString).delete();
    }
    else if (type == .file) {
      await lib.File(targetString).delete();
    }
    else if (type == .directory) {
      await lib.Directory(targetString).delete(recursive: true);
    }
    else {
      throw UnsupportedException();
    }
    return;
  }

  // #endregion

  // #region link

  static Future<Boolean> existLink(
    StoragePath target,
  ) async {
    if (target.type() != .absolute) {
      return false;
    }
    var targetString = target.emitNative();
    return await lib.FileSystemEntity.type(targetString, followLinks: false) == .link;
  }

  static Future<Void> createLink(
    StoragePath target,
    StoragePath referent,
    Boolean     isDirectory,
  ) async {
    assertTest(!await StorageHelper.exist(target));
    var targetParent = target.parent();
    if (targetParent != null && !await StorageHelper.existDirectory(targetParent)) {
      await StorageHelper.createDirectory(targetParent);
    }
    var targetString = target.emitNative();
    var referentString = referent.emitNative();
    if (SystemChecker.isWindows) {
      var hTarget = targetString.toPcwstr();
      var hReferent = referentString.toPcwstr();
      var result = lib.CreateSymbolicLink(hTarget, hReferent, lib.SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE | (!isDirectory ? 0 : lib.SYMBOLIC_LINK_FLAG_DIRECTORY));
      lib.calloc.free(hTarget);
      lib.calloc.free(hReferent);
      assertTest(result.value);
    }
    if (SystemChecker.isLinux || SystemChecker.isMacintosh || SystemChecker.isAndroid || SystemChecker.isIphone) {
      await lib.Link(targetString).create(referentString, recursive: false);
    }
    return;
  }

  // ----------------

  static Future<StoragePath> resolveLink(
    StoragePath target,
  ) async {
    assertTest(await StorageHelper.existLink(target));
    var targetString = target.emitNative();
    return .of(await lib.Link(targetString).target());
  }

  // #endregion

  // #region file

  static Future<Boolean> existFile(
    StoragePath target,
  ) async {
    if (target.type() != .absolute) {
      return false;
    }
    var targetString = target.emitNative();
    return await lib.FileSystemEntity.type(targetString, followLinks: true) == .file;
  }

  static Future<Void> createFile(
    StoragePath target,
  ) async {
    assertTest(!await StorageHelper.exist(target));
    var targetParent = target.parent();
    if (targetParent != null && !await StorageHelper.existDirectory(targetParent)) {
      await StorageHelper.createDirectory(targetParent);
    }
    var targetString = target.emitNative();
    await lib.File(targetString).create(recursive: false);
    return;
  }

  // ----------------

  static Future<Integer> sizeFile(
    StoragePath target,
  ) async {
    assertTest(await StorageHelper.existFile(target));
    var targetString = target.emitNative();
    return await lib.File(targetString).length();
  }

  static Future<Void> resizeFile(
    StoragePath target,
    Integer     size,
  ) async {
    assertTest(await StorageHelper.existFile(target));
    assertTest(size >= 0);
    var targetString = target.emitNative();
    var handler = await lib.File(targetString).open(mode: .writeOnlyAppend);
    try {
      await handler.truncate(size);
    }
    finally {
      await handler.close();
    }
    return;
  }

  // ----------------

  static Future<Void> readFile(
    StoragePath   target,
    Integer       offset,
    lib.Uint8List data,
  ) async {
    assertTest(await StorageHelper.existFile(target));
    assertTest(offset >= 0);
    var targetString = target.emitNative();
    var handler = await lib.File(targetString).open(mode: .read);
    try {
      assertTest(await handler.length() >= offset + data.length);
      await handler.setPosition(offset);
      var count = await handler.readInto(data);
      assertTest(count == data.length);
    }
    finally {
      await handler.close();
    }
    return;
  }

  static Future<Void> writeFile(
    StoragePath   target,
    Integer       offset,
    lib.Uint8List data,
  ) async {
    assertTest(await StorageHelper.existFile(target));
    assertTest(offset >= 0);
    var targetString = target.emitNative();
    var handler = await lib.File(targetString).open(mode: .writeOnlyAppend);
    try {
      assertTest(await handler.length() >= offset + data.length);
      await handler.setPosition(offset);
      await handler.writeFrom(data);
    }
    finally {
      await handler.close();
    }
    return;
  }

  // ----------------

  static Future<lib.Uint8List> readFileData(
    StoragePath target,
  ) async {
    var size = await StorageHelper.sizeFile(target);
    var data = lib.Uint8List(size);
    await StorageHelper.readFile(target, 0, data);
    return data;
  }

  static Future<Void> writeFileData(
    StoragePath   target,
    lib.Uint8List data,
  ) async {
    await StorageHelper.resizeFile(target, data.length);
    await StorageHelper.writeFile(target, 0, data);
    return;
  }

  static Future<String> readFileText(
    StoragePath target,
  ) async {
    var data = await StorageHelper.readFileData(target);
    var text = lib.utf8.decode(data);
    return text;
  }

  static Future<Void> writeFileText(
    StoragePath target,
    String      text,
  ) async {
    var data = lib.utf8.encode(text);
    await StorageHelper.writeFileData(target, data);
    return;
  }

  // #endregion

  // #region directory

  static Future<Boolean> existDirectory(
    StoragePath target,
  ) async {
    if (target.type() != .absolute) {
      return false;
    }
    var targetString = target.emitNative();
    return await lib.FileSystemEntity.type(targetString, followLinks: true) == .directory;
  }

  static Future<Void> createDirectory(
    StoragePath target,
  ) async {
    assertTest(!await StorageHelper.exist(target));
    var targetString = target.emitNative();
    await lib.Directory(targetString).create(recursive: true);
  }

  // ----------------

  static Future<List<StoragePath>> listDirectory(
    StoragePath target,
    Integer?    depth,
    Boolean     followLink,
    Boolean     allowLink,
    Boolean     allowFile,
    Boolean     allowDirectory,
  ) async {
    assertTest(await StorageHelper.existDirectory(target));
    var result = <StoragePath>[];
    var iterate = (
      Object      self,
      StoragePath currentTarget,
      StoragePath currentItem,
      Integer     currentDepth,
    ) async {
      if (depth == null || currentDepth < depth) {
        var currentTargetString = currentTarget.emitNative();
        await for (var item in lib.Directory(currentTargetString).list(recursive: false, followLinks: false)) {
          var itemType = await lib.FileSystemEntity.type(item.path, followLinks: false);
          var itemName = item.path.substring(item.path.lastIndexOf(RegExp(r'[/\\]')) + 1);
          var itemPath = currentItem.join(itemName);
          if (followLink && itemType == .link) {
            var referentType = await lib.FileSystemEntity.type(item.path, followLinks: true);
            if (referentType != .notFound) {
              itemType = referentType;
            }
          }
          if (allowLink && itemType == .link) {
            result.add(itemPath);
          }
          if (allowFile && itemType == .file) {
            result.add(itemPath);
          }
          if (allowDirectory && itemType == .directory) {
            result.add(itemPath);
          }
          if (itemType == .directory) {
            await self.as<Future<Void> Function(Object, StoragePath, StoragePath, Integer)>()(self, currentTarget.join(itemName), itemPath, currentDepth + 1);
          }
        }
      }
      return null as Void;
    };
    await iterate(iterate, target, .by(.detached), 0);
    return result;
  }

  // #endregion

  // #region shell

  static Future<StoragePath> query(
    StorageQueryType type,
  ) async {
    return (await PlatformIntegrationManager.instance.invokeQueryStorageItem(ConvertHelper.makeEnumerationToStringOfSnakeCase(type))).target;
  }

  static Future<Void> reveal(
    StoragePath target,
  ) async {
    assertTest(await StorageHelper.exist(target));
    await PlatformIntegrationManager.instance.invokeRevealStorageItem(target);
    return;
  }

  static Future<List<StoragePath>> pick(
    StoragePickType type,
    Boolean         multiply,
    StoragePath?    location,
    String?         name,
  ) async {
    if (location == null || !await StorageHelper.existDirectory(location)) {
      location = await StorageHelper.query(.userHome);
    }
    if (name == null) {
      name = '';
    }
    var target = (await PlatformIntegrationManager.instance.invokePickStorageItem(ConvertHelper.makeEnumerationToStringOfSnakeCase(type), multiply, location, name)).target;
    return target;
  }

  // ----------------

  static Future<(StoragePath, Finalizer)> temporary({
    Boolean useCache = false,
  }) async {
    var parent = await StorageHelper.query(!useCache ? .applicationTemporary : .applicationCache);
    var name = ConvertHelper.makeDateTimeToString(DateTime.now());
    var target = parent.join(name);
    var suffix = 0;
    while (await exist(target)) {
      suffix += 1;
      target = parent.join('${name}.${suffix}');
    }
    {
      await StorageHelper.createDirectory(target);
    }
    var targetFinalizer = Finalizer(() async {
      await StorageHelper.remove(target);
    });
    return (target, targetFinalizer);
  }

  // #endregion

}

File: AssistantPlus/Application/Utility/StorageHelper.cs
#pragma warning disable 0,

using Twinning.AssistantPlus;
using Windows.ApplicationModel;

namespace Twinning.AssistantPlus.Utility {

	public enum StorageQueryType {
		UserHome,
		ApplicationShared,
		ApplicationCache,
		ApplicationPackage,
		ApplicationPackagedShared,
		ApplicationPackagedCache,
	}

	public enum StoragePickType {
		LoadFile,
		LoadDirectory,
		SaveFile,
	}

	// ----------------

	public static class StorageHelper {

		#region basic

		public static async Task<Boolean> Exist(
			StoragePath target
		) {
			if (target.Type() != StoragePathType.Absolute) {
				return false;
			}
			var targetString = target.EmitNative();
			var type = new FileInfo(targetString).Attributes;
			return type != ~FileAttributes.None;
		}

		public static async Task Copy(
			StoragePath target,
			StoragePath placement,
			Boolean     followLink
		) {
			AssertTest(await StorageHelper.Exist(target));
			AssertTest(!await StorageHelper.Exist(placement));
			var placementParent = placement.Parent();
			if (placementParent != null && !await StorageHelper.ExistDirectory(placementParent)) {
				await StorageHelper.CreateDirectory(placementParent);
			}
			var targetString = target.EmitNative();
			var type = new FileInfo(targetString).Attributes;
			if (followLink && type.HasFlag(FileAttributes.ReparsePoint)) {
				var referentType = new FileInfo(targetString).ResolveLinkTarget(true).AsNotNull().Attributes;
				if (referentType != ~FileAttributes.None) {
					type = referentType;
				}
			}
			if (type.HasFlag(FileAttributes.ReparsePoint)) {
				var referent = await StorageHelper.ResolveLink(target);
				var isDirectory = new FileInfo(targetString).Attributes.HasFlag(FileAttributes.Directory);
				await StorageHelper.CreateLink(placement, referent, isDirectory);
			}
			else if (!type.HasFlag(FileAttributes.ReparsePoint) && !type.HasFlag(FileAttributes.Directory)) {
				var data = await StorageHelper.ReadFile(target);
				await StorageHelper.CreateFile(placement);
				await StorageHelper.WriteFile(placement, data);
			}
			else if (!type.HasFlag(FileAttributes.ReparsePoint) && type.HasFlag(FileAttributes.Directory)) {
				await StorageHelper.CreateDirectory(placement);
				foreach (var item in new DirectoryInfo(targetString).EnumerateFileSystemInfos("*", new EnumerationOptions() {
						AttributesToSkip = FileAttributes.None,
						BufferSize = 0,
						IgnoreInaccessible = false,
						MatchCasing = MatchCasing.PlatformDefault,
						MatchType = MatchType.Simple,
						MaxRecursionDepth = 0,
						RecurseSubdirectories = false,
						ReturnSpecialDirectories = false,
					})) {
					var itemName = item.Name;
					await StorageHelper.Copy(target.Join(itemName), placement.Join(itemName), followLink);
				}
			}
			else {
				throw new NotSupportedException();
			}
			return;
		}

		public static async Task Rename(
			StoragePath target,
			StoragePath placement
		) {
			AssertTest(await StorageHelper.Exist(target));
			AssertTest(!await StorageHelper.Exist(placement));
			var placementParent = placement.Parent();
			if (placementParent != null && !await StorageHelper.ExistDirectory(placementParent)) {
				await StorageHelper.CreateDirectory(placementParent);
			}
			var targetString = target.EmitNative();
			var placementString = placement.EmitNative();
			var type = new FileInfo(targetString).Attributes;
			if (type.HasFlag(FileAttributes.ReparsePoint)) {
				if (!type.HasFlag(FileAttributes.Directory)) {
					File.Move(targetString, placementString);
				}
				else {
					Directory.Move(targetString, placementString);
				}
			}
			else if (!type.HasFlag(FileAttributes.ReparsePoint) && !type.HasFlag(FileAttributes.Directory)) {
				File.Move(targetString, placementString);
			}
			else if (!type.HasFlag(FileAttributes.ReparsePoint) && type.HasFlag(FileAttributes.Directory)) {
				Directory.Move(targetString, placementString);
			}
			else {
				throw new NotSupportedException();
			}
			return;
		}

		public static async Task Remove(
			StoragePath target
		) {
			AssertTest(await StorageHelper.Exist(target));
			var targetString = target.EmitNative();
			var type = new FileInfo(targetString).Attributes;
			if (type.HasFlag(FileAttributes.ReparsePoint)) {
				if (!type.HasFlag(FileAttributes.Directory)) {
					File.Delete(targetString);
				}
				else {
					Directory.Delete(targetString);
				}
			}
			else if (!type.HasFlag(FileAttributes.ReparsePoint) && !type.HasFlag(FileAttributes.Directory)) {
				File.Delete(targetString);
			}
			else if (!type.HasFlag(FileAttributes.ReparsePoint) && type.HasFlag(FileAttributes.Directory)) {
				Directory.Delete(targetString, true);
			}
			else {
				throw new NotSupportedException();
			}
			return;
		}

		#endregion

		#region link

		public static async Task<Boolean> ExistLink(
			StoragePath target
		) {
			if (target.Type() != StoragePathType.Absolute) {
				return false;
			}
			var targetString = target.EmitNative();
			var type = new FileInfo(targetString).Attributes;
			return type != ~FileAttributes.None && type.HasFlag(FileAttributes.ReparsePoint);
		}

		public static async Task CreateLink(
			StoragePath target,
			StoragePath referent,
			Boolean     isDirectory
		) {
			AssertTest(!await StorageHelper.Exist(target));
			var targetParent = target.Parent();
			if (targetParent != null && !await StorageHelper.ExistDirectory(targetParent)) {
				await StorageHelper.CreateDirectory(targetParent);
			}
			var targetString = target.EmitNative();
			var referentString = referent.EmitNative();
			if (!isDirectory) {
				File.CreateSymbolicLink(targetString, referentString);
			}
			else {
				Directory.CreateSymbolicLink(targetString, referentString);
			}
			return;
		}

		// ----------------

		public static async Task<StoragePath> ResolveLink(
			StoragePath target
		) {
			AssertTest(await StorageHelper.ExistLink(target));
			var targetString = target.EmitNative();
			return new (new FileInfo(targetString).LinkTarget.AsNotNull());
		}

		#endregion

		#region file

		public static async Task<Boolean> ExistFile(
			StoragePath target
		) {
			if (target.Type() != StoragePathType.Absolute) {
				return false;
			}
			var targetString = target.EmitNative();
			var type = new FileInfo(targetString).Attributes;
			if (type != ~FileAttributes.None && type.HasFlag(FileAttributes.ReparsePoint)) {
				type = new FileInfo(targetString).ResolveLinkTarget(true).AsNotNull().Attributes;
			}
			return type != ~FileAttributes.None && !type.HasFlag(FileAttributes.Directory);
		}

		public static async Task CreateFile(
			StoragePath target
		) {
			AssertTest(!await StorageHelper.Exist(target));
			var targetParent = target.Parent();
			if (targetParent != null && !await StorageHelper.ExistDirectory(targetParent)) {
				await StorageHelper.CreateDirectory(targetParent);
			}
			var targetString = target.EmitNative();
			File.Create(targetString).Close();
			return;
		}

		// ----------------

		public static async Task<Integer> SizeFile(
			StoragePath target
		) {
			AssertTest(await StorageHelper.ExistFile(target));
			var targetString = target.EmitNative();
			return new FileInfo(targetString).Length;
		}

		// ----------------

		public static async Task<Byte[]> ReadFile(
			StoragePath target
		) {
			AssertTest(await StorageHelper.ExistFile(target));
			var targetString = target.EmitNative();
			return await File.ReadAllBytesAsync(targetString);
		}

		public static async Task WriteFile(
			StoragePath target,
			Byte[]      data
		) {
			AssertTest(await StorageHelper.ExistFile(target));
			var targetString = target.EmitNative();
			await File.WriteAllBytesAsync(targetString, data);
			return;
		}

		// ----------------

		public static async Task<String> ReadFileText(
			StoragePath target
		) {
			AssertTest(await StorageHelper.ExistFile(target));
			var targetString = target.EmitNative();
			return await File.ReadAllTextAsync(targetString);
		}

		public static async Task WriteFileText(
			StoragePath target,
			String      text
		) {
			AssertTest(await StorageHelper.ExistFile(target));
			var targetString = target.EmitNative();
			await File.WriteAllTextAsync(targetString, text);
			return;
		}

		// ----------------

		public static async Task<Byte[]> ReadFileLimited(
			StoragePath target,
			Size        limit
		) {
			AssertTest(await StorageHelper.ExistFile(target));
			var targetString = target.EmitNative();
			await using var stream = File.OpenRead(targetString);
			var size = Math.Min(stream.Length.CastPrimitive<Size>(), limit);
			var data = new Byte[size];
			var sizeActual = await stream.ReadAsync(data, 0, size);
			AssertTest(sizeActual == size);
			return data;
		}

		#endregion

		#region directory

		public static async Task<Boolean> ExistDirectory(
			StoragePath target
		) {
			if (target.Type() != StoragePathType.Absolute) {
				return false;
			}
			var targetString = target.EmitNative();
			var type = new FileInfo(targetString).Attributes;
			if (type != ~FileAttributes.None && type.HasFlag(FileAttributes.ReparsePoint)) {
				type = new FileInfo(targetString).ResolveLinkTarget(true).AsNotNull().Attributes;
			}
			return type != ~FileAttributes.None && type.HasFlag(FileAttributes.Directory);
		}

		public static async Task CreateDirectory(
			StoragePath target
		) {
			AssertTest(!await StorageHelper.Exist(target));
			var targetString = target.EmitNative();
			Directory.CreateDirectory(targetString);
			return;
		}

		// ----------------

		public static async Task<List<StoragePath>> ListDirectory(
			StoragePath target,
			Size?       depth,
			Boolean     followLink,
			Boolean     allowLink,
			Boolean     allowFile,
			Boolean     allowDirectory
		) {
			AssertTest(await StorageHelper.ExistDirectory(target));
			var result = new List<StoragePath>();
			async Task iterate(
				StoragePath currentTarget,
				StoragePath currentItem,
				Integer     currentDepth
			) {
				if (depth == null || currentDepth < depth) {
					var currentTargetString = currentTarget.EmitNative();
					foreach (var item in new DirectoryInfo(currentTargetString).EnumerateFileSystemInfos("*", new EnumerationOptions() {
							AttributesToSkip = FileAttributes.None,
							BufferSize = 0,
							IgnoreInaccessible = false,
							MatchCasing = MatchCasing.PlatformDefault,
							MatchType = MatchType.Simple,
							MaxRecursionDepth = 0,
							RecurseSubdirectories = false,
							ReturnSpecialDirectories = false,
						})) {
						var itemType = item.Attributes;
						var itemName = item.Name;
						var itemPath = currentItem.Join(itemName);
						if (followLink && itemType.HasFlag(FileAttributes.ReparsePoint)) {
							var referentType = item.ResolveLinkTarget(true).AsNotNull().Attributes;
							if (referentType != ~FileAttributes.None) {
								itemType = referentType;
							}
						}
						if (allowLink && itemType.HasFlag(FileAttributes.ReparsePoint)) {
							result.Add(itemPath);
						}
						if (allowFile && !itemType.HasFlag(FileAttributes.ReparsePoint) && !itemType.HasFlag(FileAttributes.Directory)) {
							result.Add(itemPath);
						}
						if (allowDirectory && !itemType.HasFlag(FileAttributes.ReparsePoint) && itemType.HasFlag(FileAttributes.Directory)) {
							result.Add(itemPath);
						}
						if (!itemType.HasFlag(FileAttributes.ReparsePoint) && itemType.HasFlag(FileAttributes.Directory)) {
							await iterate(currentTarget.Join(itemName), itemPath, currentDepth + 1);
						}
					}
				}
				return;
			}
			await iterate(target, new (StoragePathType.Detached), 0);
			return result;
		}

		#endregion

		#region shell

		public static async Task<StoragePath> Query(
			StorageQueryType type
		) {
			var path = null as StoragePath;
			switch (type) {
				case StorageQueryType.UserHome: {
					path = new ($"{Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)}");
					break;
				}
				case StorageQueryType.ApplicationShared: {
					path = new ($"{Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData)}\\{ApplicationInformation.Identifier}");
					break;
				}
				case StorageQueryType.ApplicationCache: {
					path = new ($"{Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData)}\\{ApplicationInformation.Identifier}\\cache");
					break;
				}
				case StorageQueryType.ApplicationPackage: {
					path = new ($"{Package.Current.InstalledPath}");
					break;
				}
				case StorageQueryType.ApplicationPackagedShared: {
					path = new ($"{Windows.Storage.ApplicationData.Current.LocalFolder.Path}");
					break;
				}
				case StorageQueryType.ApplicationPackagedCache: {
					path = new ($"{Windows.Storage.ApplicationData.Current.LocalFolder.Path}\\cache");
					break;
				}
				default: throw new UnreachableException();
			}
			return path;
		}

		public static async Task Reveal(
			StoragePath target
		) {
			AssertTest(await StorageHelper.Exist(target));
			var targetString = target.EmitNative();
			var result = Win32.PInvoke.ShellExecute(new (IntegerSN.Zero), "open", $"file://{targetString}", null, null, Win32.UI.WindowsAndMessaging.SHOW_WINDOW_CMD.SW_SHOWNORMAL);
			AssertTest(result >= 32);
			return;
		}

		public static async Task<StoragePath?> Pick(
			StoragePickType type,
			StoragePath?    location,
			String?         name,
			Window          host
		) {
			if (location == null || !await StorageHelper.ExistDirectory(location)) {
				location = await StorageHelper.Query(StorageQueryType.UserHome);
			}
			if (name == null) {
				name = "";
			}
			var target = null as StoragePath;
			await Task.Run(() => {
				unsafe {
					Win32.PInvoke.CoCreateInstance<Win32.UI.Shell.IFileDialog>(
						type switch {
							StoragePickType.LoadFile      => typeof(Win32.UI.Shell.FileOpenDialog).GUID,
							StoragePickType.LoadDirectory => typeof(Win32.UI.Shell.FileOpenDialog).GUID,
							StoragePickType.SaveFile      => typeof(Win32.UI.Shell.FileSaveDialog).GUID,
							_                             => throw new UnreachableException(),
						},
						null,
						Win32.System.Com.CLSCTX.CLSCTX_INPROC_SERVER,
						out var dialog
					).ThrowOnFailure();
					using var dialogFinalizer = new Finalizer(async () => {
						dialog->Release();
					});
					var option = (Win32.UI.Shell.FILEOPENDIALOGOPTIONS)0;
					option |= Win32.UI.Shell.FILEOPENDIALOGOPTIONS.FOS_NOCHANGEDIR;
					option |= Win32.UI.Shell.FILEOPENDIALOGOPTIONS.FOS_NODEREFERENCELINKS;
					option |= Win32.UI.Shell.FILEOPENDIALOGOPTIONS.FOS_FORCESHOWHIDDEN;
					if (type == StoragePickType.LoadFile || type == StoragePickType.LoadDirectory) {
						option |= Win32.UI.Shell.FILEOPENDIALOGOPTIONS.FOS_PATHMUSTEXIST;
						option |= Win32.UI.Shell.FILEOPENDIALOGOPTIONS.FOS_FILEMUSTEXIST;
					}
					if (type == StoragePickType.LoadDirectory) {
						option |= Win32.UI.Shell.FILEOPENDIALOGOPTIONS.FOS_PICKFOLDERS;
					}
					if (type == StoragePickType.SaveFile) {
						option |= Win32.UI.Shell.FILEOPENDIALOGOPTIONS.FOS_PATHMUSTEXIST;
						option |= Win32.UI.Shell.FILEOPENDIALOGOPTIONS.FOS_OVERWRITEPROMPT;
						option |= Win32.UI.Shell.FILEOPENDIALOGOPTIONS.FOS_NOREADONLYRETURN;
					}
					dialog->SetOptions(option).ThrowOnFailure();
					Win32.PInvoke.SHCreateItemFromParsingName(
						location.EmitNative(),
						null,
						typeof(Win32.UI.Shell.IShellItem).GUID,
						out var locationItem
					).ThrowOnFailure();
					using var locationItemFinalizer = new Finalizer(async () => {
						((Win32.UI.Shell.IShellItem*)locationItem)->Release();
					});
					dialog->SetFolder((Win32.UI.Shell.IShellItem*)locationItem).ThrowOnFailure();
					if (type == StoragePickType.SaveFile) {
						dialog->SetFileName(name).ThrowOnFailure();
					}
					var dialogResult = dialog->Show(new (WindowHelper.GetHandle(host)));
					if (dialogResult != Win32.PInvoke.HRESULT_FROM_WIN32(Win32.Foundation.WIN32_ERROR.ERROR_CANCELLED)) {
						dialogResult.ThrowOnFailure();
						var targetItem = default(Win32.UI.Shell.IShellItem*);
						dialog->GetResult(&targetItem).ThrowOnFailure();
						var targetItemPointer = &targetItem;
						using var targetItemFinalizer = new Finalizer(async () => {
							(*targetItemPointer)->Release();
						});
						targetItem->GetDisplayName(Win32.UI.Shell.SIGDN.SIGDN_FILESYSPATH, out var targetPath).ThrowOnFailure();
						using var targetPathFinalizer = new Finalizer(async () => {
							Win32.PInvoke.CoTaskMemFree(targetPath.Value);
						});
						target = new (targetPath.ToString());
					}
				}
			});
			return target;
		}

		// ----------------

		public static async Task<StoragePath> Temporary(
		) {
			var parent = App.Instance.CacheDirectory;
			var name = DateTime.Now.Ticks.ToString(CultureInfo.InvariantCulture);
			var result = parent.Join(name);
			var suffix = 0;
			while (await StorageHelper.Exist(result)) {
				suffix += 1;
				result = parent.Join($"{name}.{suffix}");
			}
			return result;
		}

		public static async Task Trash(
			StoragePath target
		) {
			AssertTest(await StorageHelper.Exist(target));
			var targetString = target.EmitNative();
			if (await StorageHelper.ExistFile(target)) {
				Microsoft.VisualBasic.FileIO.FileSystem.DeleteFile(targetString, Microsoft.VisualBasic.FileIO.UIOption.OnlyErrorDialogs, Microsoft.VisualBasic.FileIO.RecycleOption.SendToRecycleBin, Microsoft.VisualBasic.FileIO.UICancelOption.ThrowException);
			}
			if (await StorageHelper.ExistDirectory(target)) {
				Microsoft.VisualBasic.FileIO.FileSystem.DeleteDirectory(targetString, Microsoft.VisualBasic.FileIO.UIOption.OnlyErrorDialogs, Microsoft.VisualBasic.FileIO.RecycleOption.SendToRecycleBin, Microsoft.VisualBasic.FileIO.UICancelOption.ThrowException);
			}
			return;
		}

		#endregion

		#region miscellaneous

		private static readonly Character[] InvalidPathNameCharacter = Path.GetInvalidFileNameChars();

		public static Boolean CheckName(
			String name
		) {
			if (name.IsEmpty()) {
				return false;
			}
			if (name[0] == ' ' || name[^1] == ' ' || name[^1] == '.') {
				return false;
			}
			return name.All((value) => !StorageHelper.InvalidPathNameCharacter.Contains(value));
		}

		// ----------------

		public static StoragePath ResolveLongPath(
			String source
		) {
			var destinationLength = Win32.PInvoke.GetLongPathName(source, []);
			var destination = new Character[destinationLength];
			var destinationLengthCheck = Win32.PInvoke.GetLongPathName(source, destination.AsSpan());
			AssertTest(destinationLengthCheck == destinationLength - 1);
			return new (new String(destination.AsSpan(0, destinationLength.CastPrimitive<Size>() - 1)));
		}

		#endregion

	}

}
